### PR TITLE
[build] download r8 from maven, instead of building from source

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: https://www.reddit.com/r/androiddev/comments/a4wpuk/for_anyone_wondering_about_todays_failed_builds/
Context: https://issuetracker.google.com/issues/120759347
Context: https://mvnrepository.com/artifact/com.android.tools/r8/1.2.52

We are currently building r8 from source, which has been failing on
occasion:

    04:31:34 _BuildR8:
    04:31:34   python tools/gradle.py r8 --stacktrace --no-daemon
    04:31:34   Running: download_from_google_storage -n -b r8-deps -u -s /mnt/jenkins/workspace/xamarin-anroid-linux-pr-builder/xamarin-android/external/r8/third_party/gradle/gradle.tar.gz.sha1
    04:31:41   0> Downloading /mnt/jenkins/workspace/xamarin-anroid-linux-pr-builder/xamarin-android/external/r8/third_party/gradle/gradle.tar.gz...
    04:31:41   0> Extracting 196 entries from /mnt/jenkins/workspace/xamarin-anroid-linux-pr-builder/xamarin-android/external/r8/third_party/gradle/gradle.tar.gz to /mnt/jenkins/workspace/xamarin-anroid-linux-pr-builder/xamarin-android/external/r8/third_party/gradle/gradle
    04:31:41   Downloading 1 files took 7.041677 second(s)
    04:31:41   Running: /mnt/jenkins/workspace/xamarin-anroid-linux-pr-builder/xamarin-android/external/r8/third_party/gradle/gradle/bin/gradle r8 --stacktrace --no-daemon
    04:31:44   :buildSrc:clean UP-TO-DATE
    04:31:50   :buildSrc:compileJava
    04:31:50   :buildSrc:compileGroovy NO-SOURCE
    04:31:50   :buildSrc:processResources NO-SOURCE
    04:31:50   :buildSrc:classes
    04:31:50   :buildSrc:jar
    04:31:50   :buildSrc:assemble
    04:31:50   :buildSrc:compileTestJava NO-SOURCE
    04:31:50   :buildSrc:compileTestGroovy NO-SOURCE
    04:31:50   :buildSrc:processTestResources NO-SOURCE
    04:31:50   :buildSrc:testClasses UP-TO-DATE
    04:31:50   :buildSrc:test NO-SOURCE
    04:31:50   :buildSrc:check UP-TO-DATE
    04:31:50   :buildSrc:build
    04:31:52
    04:31:52   FAILURE: Build failed with an exception.
    04:31:52
    04:31:52   * What went wrong:
    04:31:52   A problem occurred configuring root project 'r8'.
    04:31:52   > Could not resolve all dependencies for configuration ':classpath'.
    04:31:52      > Could not find com.android.tools:common:25.3.1.
    04:31:52        Searched in the following locations:
    04:31:52            https://repo1.maven.org/maven2/com/android/tools/common/25.3.1/common-25.3.1.pom
    04:31:52            https://repo1.maven.org/maven2/com/android/tools/common/25.3.1/common-25.3.1.jar
    04:31:52            https://jcenter.bintray.com/com/android/tools/common/25.3.1/common-25.3.1.pom
    04:31:52            https://jcenter.bintray.com/com/android/tools/common/25.3.1/common-25.3.1.jar
    04:31:52            https://plugins.gradle.org/m2/com/android/tools/common/25.3.1/common-25.3.1.pom
    04:31:52            https://plugins.gradle.org/m2/com/android/tools/common/25.3.1/common-25.3.1.jar
    04:31:52        Required by:
    04:31:52            project : > com.cookpad.android.licensetools:license-tools-plugin:0.23.0 > com.android.tools.build:gradle:2.3.1 > com.android.tools.build:gradle-core:2.3.1 > com.android.tools.build:builder:2.3.1
    04:31:52            project : > com.cookpad.android.licensetools:license-tools-plugin:0.23.0 > com.android.tools.build:gradle:2.3.1 > com.android.tools.build:gradle-core:2.3.1 > com.android.tools.build:builder:2.3.1 > com.android.tools.build:manifest-merger:25.3.1
    04:31:52            project : > com.cookpad.android.licensetools:license-tools-plugin:0.23.0 > com.android.tools.build:gradle:2.3.1 > com.android.tools.build:gradle-core:2.3.1 > com.android.tools.build:builder:2.3.1 > com.android.tools.ddms:ddmlib:25.3.1
    04:31:52            project : > com.cookpad.android.licensetools:license-tools-plugin:0.23.0 > com.android.tools.build:gradle:2.3.1 > com.android.tools.build:gradle-core:2.3.1 > com.android.tools.build:builder:2.3.1 > com.android.tools.analytics-library:shared:25.3.1
    04:31:52            project : > com.cookpad.android.licensetools:license-tools-plugin:0.23.0 > com.android.tools.build:gradle:2.3.1 > com.android.tools.build:gradle-core:2.3.1 > com.android.tools.build:builder:2.3.1 > com.android.tools.analytics-library:tracker:25.3.1
    04:31:52            project : > com.cookpad.android.licensetools:license-tools-plugin:0.23.0 > com.android.tools.build:gradle:2.3.1 > com.android.tools.build:gradle-core:2.3.1 > com.android.tools.build:builder:2.3.1 > com.android.tools:sdklib:25.3.1 > com.android.tools.layoutlib:layoutlib-api:25.3.1
    04:31:52            project : > com.cookpad.android.licensetools:license-tools-plugin:0.23.0 > com.android.tools.build:gradle:2.3.1 > com.android.tools.build:gradle-core:2.3.1 > com.android.tools.build:builder:2.3.1 > com.android.tools:sdklib:25.3.1 > com.android.tools:dvlib:25.3.1
    04:31:52            project : > com.cookpad.android.licensetools:license-tools-plugin:0.23.0 > com.android.tools.build:gradle:2.3.1 > com.android.tools.build:gradle-core:2.3.1 > com.android.tools.build:builder:2.3.1 > com.android.tools:sdklib:25.3.1 > com.android.tools:repository:25.3.1
    04:31:52
    04:31:52   * Try:
    04:31:52   Run with --info or --debug option to get more log output.
    04:31:52
    04:31:52   * Exception is:
    04:31:52   org.gradle.api.ProjectConfigurationException: A problem occurred configuring root project 'r8'.

I have seen this on Linux and Windows PR builds, which we think is
caused by Google removing `com.android.tools:common:25.3.1` from
maven. Some of our local machines are still working, likely due to
local caching by gradle. I don't think r8 as-is will build on a clean
machine any longer!

In either case, we've been wanting to use `r8.jar` from maven. It is a
little concerning that Google has pulled packages from there, but
something that is working *now* is better than not working at all.

Since we no longer build `r8` from source, I was able to delete a lot
of code. We don't need to download Chromium's `depot_tools` any more.

`r8` from maven also lists a different license:

https://opensource.org/licenses/BSD-3-Clause

I copied the contents to `src/r8/LICENSE` and moved `r8.tpnitems` to a
more appropriate location.

## gradle ##

When downloading the file directly from maven, if you tried to use it,
you would run into:

    _CompileToDalvikWithD8:
        C:\Program Files (x86)\Java\jdk1.8.0_161\bin\java.exe -Xmx1G -classpath C:\Users\myuser\Desktop\Git\xamarin-android\bin\Debug\lib\xamarin.android\xbuild\Xamarin\Android\r8.jar com.android.tools.r8.D8 --debug --min-api 21 --output obj\Debug\android\b
        in --lib C:\Users\myuser\android-toolchain\sdk\platforms\android-28\android.jar obj\Debug\android\bin\classes.zip C:\Users\myuser\Desktop\Git\xamarin-android\bin\Debug\lib\xamarin.android\xbuild-frameworks\MonoAndroid\v9.0\mono.android.jar
        Exception in thread "main" java.lang.NoClassDefFoundError: com/google/common/collect/ImmutableList
                at com.android.tools.r8.D8CommandParser$OrderedClassFileResourceProvider$Builder.<init>(D8CommandParser.java:30)
                at com.android.tools.r8.D8CommandParser$OrderedClassFileResourceProvider.builder(D8CommandParser.java:58)
                at com.android.tools.r8.D8CommandParser.parse(D8CommandParser.java:143)
                at com.android.tools.r8.D8CommandParser.parse(D8CommandParser.java:120)
                at com.android.tools.r8.D8Command.parse(D8Command.java:177)
                at com.android.tools.r8.D8.run(D8.java:98)
                at com.android.tools.r8.D8.lambda$main$3(D8.java:122)
                at com.android.tools.r8.utils.ExceptionUtils.withMainProgramHandler(ExceptionUtils.java:75)
                at com.android.tools.r8.D8.main(D8.java:122)
        Caused by: java.lang.ClassNotFoundException: com.google.common.collect.ImmutableList
                at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
                at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
                at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:338)
                at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
                ... 9 more

r8 has ~9 dependencies, which are not included in `r8.jar` itself.

So to workaround this:

1. Created a "dummy" java project using gradle.
2. Setup a dependency on:
   `group: 'com.android.tools', name: 'r8', version: '1.2.52'`
3. Setup the `jar` gradle task to zip all dependencies into a
   self-contained jar file:

```groovy
from {
    configurations.compile.collect { it.isDirectory() ? it : zipTree(it) }
}
```

When running:

    .\gradlew jar

A `build\libs\r8.jar` file is produced, containing all dependencies.

I also consolidated our existing use of gradle:

- Created a new `build-tools\gradle` directory.
- Keep one copy of `gradlew`, `gradlew.bat`, and `gradle-wrapper.jar`.
- Added new properties to `Configuration.props`: `$(GradleHome)`,
  `$(GradleWPath)`, and `$(GradleArgs)`.
- Places that call gradle now uses these new properties and set
  `APP_HOME=$(GradleHome)`.

### CODE REVIEW

**Comment by jonathanpeppers** [comment]
Ugh ok:
```
_CompileToDalvikWithD8:
  C:\Program Files (x86)\Java\jdk1.8.0_161\bin\java.exe -Xmx1G -classpath C:\Users\myuser\Desktop\Git\xamarin-android\bin\Debug\lib\xamarin.android\xbuild\Xamarin\Android\r8.jar com.android.tools.r8.D8 --debug --min-api 21 --output obj\Debug\android\b
  in --lib C:\Users\myuser\android-toolchain\sdk\platforms\android-28\android.jar obj\Debug\android\bin\classes.zip C:\Users\myuser\Desktop\Git\xamarin-android\bin\Debug\lib\xamarin.android\xbuild-frameworks\MonoAndroid\v9.0\mono.android.jar
  Exception in thread "main" java.lang.NoClassDefFoundError: com/google/common/collect/ImmutableList
        at com.android.tools.r8.D8CommandParser$OrderedClassFileResourceProvider$Builder.<init>(D8CommandParser.java:30)
        at com.android.tools.r8.D8CommandParser$OrderedClassFileResourceProvider.builder(D8CommandParser.java:58)
        at com.android.tools.r8.D8CommandParser.parse(D8CommandParser.java:143)
        at com.android.tools.r8.D8CommandParser.parse(D8CommandParser.java:120)
        at com.android.tools.r8.D8Command.parse(D8Command.java:177)
        at com.android.tools.r8.D8.run(D8.java:98)
        at com.android.tools.r8.D8.lambda$main$3(D8.java:122)
        at com.android.tools.r8.utils.ExceptionUtils.withMainProgramHandler(ExceptionUtils.java:75)
        at com.android.tools.r8.D8.main(D8.java:122)
  Caused by: java.lang.ClassNotFoundException: com.google.common.collect.ImmutableList
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:338)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 9 more
```

Dependencies are:
image

So not sure what to do... Do we want to bundle all this stuff???

**Comment by jonathanpeppers** [comment]
Ok, Google might just be fixing our original build failure: https://issuetracker.google.com/issues/120759347

> Hey everyone,
>
> We understand this is a really painful situation right now for a lot of developers.  Removing packages from jCenter which are not replicated in Google's maven repo was an accident and a product of a miscommunication!  We're working on getting everything back into place. 
>
> Please try and be patient with us and let's try and reserve comments on this thread for updates on the situation.   If you're affected by this issue and want to be heard, please click "Me too!" at the top of the page as a way of 'upvoting' this issue.
>
> - Sam

So we still have an exploration of using `r8.jar` from maven, and it seems to me we would still have to make a "dummy" gradle build that merges all these dependencies into a final `r8.jar`. gradle would also handle the downloads. Is that the best plan if we wanted to do this?

Example of using gradle to merge jars: https://discuss.gradle.org/t/how-to-include-dependencies-in-jar/19571/5

**Comment by jonpryor** [comment]
This would add a *second* `gradle-wrapper.jar`; the current one is at:

https://github.com/xamarin/xamarin-android/tree/3f743678d3e7c19fc6f6e8f6928bfe4ae7d55ff0/tests/CodeGen-Binding/Xamarin.Android.LibraryProjectZip-LibBinding/java/JavaLib/gradle/wrapper

Is there any way to unify these?

**Comment by jonpryor** [comment]
@jonathanpeppers asked:

> Is that the best plan if we wanted to do this?

¯\_(ツ)_/¯

What is the size of the merged `.jar` that this PR produces -- presumably containing [all dependencies](https://github.com/xamarin/xamarin-android/pull/2519#issuecomment-446276689) -- vs. what we *currently* produce?  Is there a significant size difference?

**Comment by jonpryor** [comment]
Meaning that the `r8.jar` from this PR is ~827KB *smaller* than what we currently have.

I wonder how worrying that should be...?

If this PR works, we should probably go for it then, as it's a size reduction.

File: build-tools/gradle/gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,6 @@
-#Thu Dec 14 15:29:43 EST 2017
+#Tue Dec 11 14:50:37 CST 2018
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.4-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.10-all.zip
```
Comment (jonathanpeppers) [inline build-tools/gradle/gradle/wrapper/gradle-wrapper.properties]: I used the latest IntelliJ to create the new project, it used gradle 4.10.

File: build-tools/gradle/gradlew [hunk 2]
```diff
@@ -155,7 +155,7 @@ if $cygwin ; then
 fi
 
 # Escape application args
-save ( ) {
+save () {
     for i do printf %s\\n "$i" | sed "s/'/'\\\\''/g;1s/^/'/;\$s/\$/' \\\\/" ; done
     echo " "
 }
```
Comment (jonathanpeppers) [inline build-tools/gradle/gradlew]: Hmm, I guess the latest wrapper script from the new project changed formatting.

**Pull request merged**